### ISSUE TITLE
Thorium not exposing MathML to NVDA and probably Jaws

### ISSUE THREAD

**Issue by GeorgeKerscher** [edrlab/thorium-reader#1602]
Using Thorium v1.7.4-alpha.1.1561269451 on Windows, it seems that the MathML is not exposed.

Using NVDA 2021.3, which was released in the past week, I am inside a MathML expression in epubtest.org Math test book. I hit NVDA key + alt +m and it says not math.

I tried in Jaws, and it did not recognize it as math as well.

I tested with 1.7.3 and the MathML is read. Noelia reported this to me and asked that I submit an issue.

**Comment by nvdaes** [comment]
Thanks @GeorgeKerscher, I wanted to test mathematics with the alpha version of Thorium, but I switched to 1.7.3 since it support mathml with NVDA (100%).
Best,
Noelia

**Comment by nvdaes** [comment]
Hello, without Mathjax, Mathml is not exposed to MathPlayer. With Mathjax, NVDA detets an application, but at least with default settings it cannot navigated properly.
Thanks

**Comment by danielweck** [comment]
Hello, I tested with VoiceOver on the latest version of MacOS.  When I navigate to the MathML parts rendered with MathJax, I hear the arial-label text generated dynamically by MathJax (i.e. not authored ahead of time). This seems correct to me as the aural rendition more or less matches the hard-coded authored alt text for visual math images.
I am using `Accessibility-Tests-Mathematics-v1.1.1.epub` to run my tests.

**Comment by GeorgeKerscher** [comment]
I just tested with Thorium v1.7.4-alpha.1.1584412998 and using both Jaws and NVDA I could not read the MathML. I tested with MathJax turned and off. In NVDA, it read the alt-text inside the MathML, and that was it. When I tested the same math 1.1 title unpacked in Chrome, NVDA read the MathML.

**Comment by danielweck** [comment]
What do you mean by "read the MathML"? Do you mean when a screen reader enters / navigates the deep MathML markup structure even though there is an alttext / aria-label at the root? Note that when MathJax is activated in Thorium (or directly inside authored scripted content documents, for that matter), MathJax preserves the MathML somewhere deep in the DOM fragment but it exposes its own SVG or HTML visual representation, and it generates a speakable alt text detectable at the fragment root. In my tests, both the screen reader synthetic speech and Thorium's built-in TTS narrate this alt text. If there is no such alt text, then I would expect the screen reader to parse the deep MathML or MathJax in order to extract the available text (which usually sounds unintelligible).
Now, I am aware that MathJax optionally implements its own navigable deep MathML structure, but this is an advanced interactive case so I do not believe this is what you are referring to?
Thank you for clarifying.

**Comment by GeorgeKerscher** [comment]
Hi,

 

With NVDA, I have an add-in called Access8Math. When in MathML, it provides a user interface that allows precise navigation of the math and moving up and down in the structures. Using 1.7.3, I can use this, but inThorium v1.7.4-alpha.1.1584412998 with  or without MathJax active, I cannot get into it. Also the NVDA + alt + m says, not math.

 

Best

George

**Comment by danielweck** [comment]
Here is what MathJax documentation says:

* http://docs.mathjax.org/en/v3.2-latest/basic/accessibility.html#screen-reader-support
* http://docs.mathjax.org/en/v3.2-latest/options/accessibility.html#assistive-mml-options
* http://docs.mathjax.org/en/v3.2-latest/web/components/accessibility.html#a11y-assistive-mml

**Comment by danielweck** [comment]
Hello, the latest automated build of Thorium will include a workaround for this. In a nutshell: when a screen reader is detected, MathJax does not generate its own accessible MathML description anymore, which results in the previous behaviour which allowed screen readers to connect to the underlying structure and work out their own rendition (e.g. JAWS or NVDA + plugin). Conversely, when no screen reader is detected, we allow MathJax to generate its own MathML label which Thorium speaks via TTS during its own read aloud experience.
